modules/renter/repair.go

```diff
@@ -87,9 +86,19 @@ func (f *file) repair(chunkIndex uint64, missingPieces []uint64, r io.ReaderAt,
 			})
 			f.contracts[host.ContractID()] = contract
 			f.mu.Unlock()
+			errChan <- nil
 		}(missingPieces[i], hosts[i])
 	}
-	wg.Wait()
+	var errs []error
+	for i := 0; i < numPieces; i++ {
+		err := <-errChan
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if len(errs) == numPieces {
+		return fmt.Errorf("could not upload any sectors:\n%v", build.JoinErrors(errs, "\n"))
```

Comment:
I'm thinking `proto` should have its own error type:
```go
type Error struct  {
    error
    transient bool
}

func IsTransient(err error) {
    e, ok := err.(*Error)
    return ok && e.transient
}
```
Then the renter can explicitly check for transient errors, and it will know not to retry uploading after a non-transient error (e.g. `contract has ended`). Transient errors should be handled with exponential backoff, as you suggested.
I suppose the utility of this depends on how many non-transient errors there are. I think it's worth it if there's more than two.